File: leetcode/src/main/java/com/giwankim/leetcode/BinaryTreeLevelOrderTraversal.java
package com.giwankim.leetcode;

import com.giwankim.leetcode.support.TreeNode;
import java.util.*;

public class BinaryTreeLevelOrderTraversal {
  /**
   * @implNote Time {@code O(n)}, space {@code O(n)}.
   */
  public List<List<Integer>> levelOrder(TreeNode root) {
    if (root == null) {
      return Collections.emptyList();
    }
    List<List<Integer>> result = new ArrayList<>();
    Queue<TreeNode> queue = new ArrayDeque<>();
    queue.offer(root);
    while (!queue.isEmpty()) {
      int size = queue.size();
      List<Integer> level = new ArrayList<>();
      while (size-- > 0) {
        TreeNode node = queue.poll();
        level.add(node.val);
        if (node.left != null) {
          queue.offer(node.left);
        }
        if (node.right != null) {
          queue.offer(node.right);
        }
      }
      result.add(level);
    }
    return result;
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/BinaryTreeLevelOrderTraversal.kt
package com.giwankim.leetcode

import com.giwankim.leetcode.support.TreeNode
import java.util.LinkedList
import java.util.Queue

class BinaryTreeLevelOrderTraversal {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        if (root == null) {
            return emptyList()
        }
        val result = mutableListOf<List<Int>>()
        val queue: Queue<TreeNode> = LinkedList<TreeNode>()
        queue.add(root)
        while (queue.isNotEmpty()) {
            val level = mutableListOf<Int>()
            val size = queue.size
            for (i in 0 until size) {
                val node = queue.poll()
                level.add(node.`val`)
                if (node.left != null) {
                    queue.add(node.left)
                }
                if (node.right != null) {
                    queue.add(node.right)
                }
            }
            result.add(level)
        }
        return result
    }
}
